sfge/src/main/java/com/salesforce/graph/ApexPath.java

```diff
@@ -193,6 +186,38 @@ public ApexPath deepClone() {
         return new ApexPath(this);
     }
 
+    /**
+     * Get every subpath contained within this path.
+     *
+     * @param deep - Also include subpaths of subpaths
+     */
+    public List<ApexPath> getAllSubpaths(boolean deep) {
+        List<ApexPath> results = new ArrayList<>();
+        if (instanceInitializationPath != null) {
+            results.add(instanceInitializationPath);
+        }
+        if (constructorPath != null) {
+            results.add(constructorPath);
+        }
+        List<Collectible<ApexPath>> collectiblePaths =
+                new ArrayList<>(newInstanceToInitializationPath.values());
+        collectiblePaths.addAll(staticClassNameToInitializationPath.values());
+        for (Collectible<ApexPath> pathCollectible : collectiblePaths) {
+            if (pathCollectible.getCollectible() != null) {
+                results.add(pathCollectible.getCollectible());
+            }
+        }
+        results.addAll(invocableVertexToPath.values());
+        if (deep) {
+            int i = 0;
+            while (i < results.size()) {
+                results.addAll(results.get(i).getAllSubpaths(true));
+                i++;
+            }
+        }
```

Comment:
The more I think about this, the more it feels like we could avoid this recursion - we already perform two recursions before we reach this point:
- the first time to identify the paths, 
- and the second time to walk the identified paths while executing the rules. 

We might as well collect the same information in the first recursion instead of performing recursion a third time. 

Even though we aren't trying to prematurely optimize performance, it somehow feels like we're setting ourselves up for a revisit.

Let's talk about this when you're back. In the meantime, I'll also think about this to see if we even have the possibility of an alternative. I think most of the code you have here will still be needed no matter which approach we end up with.

Reply:
Discovered something helpful - have a look at [ApexPathExpanderConfig](https://github.com/forcedotcom/sfdx-scanner/blob/dev/sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpanderConfig.java) on how it's used as a hook mechanism to inject things into path expansion logic. It gets populated from [PathBasedRuleRunner](https://github.com/forcedotcom/sfdx-scanner/blob/46115496b404556983f946f3348df1e38dabf063/sfge/src/main/java/com/salesforce/rules/PathBasedRuleRunner.java#L199).